Model specific microservices

mruwnik opened the pull request:
The models can be accessed over HTTP, below are curl calls:
* reader model (for duplicates)
 `curl -X POST -H "Content-Type: application/json" -d '{"query": "generate something for this", "context": "asdasdasd"}' -H "Authorization: Bearer $(gcloud auth print-identity-token)" https://reader-model-t6p37v2uia-uw.a.run.app/`
* retriever model (returns encodings)
`curl -X POST -H "Content-Type: application/json" -d '{"query": "generate something for this", "context": "asdasdasd"}' -H "Authorization: Bearer $(gcloud auth print-identity-token)" https://retriever-model-t6p37v2uia-uw.a.run.app`
* literature search (returns encodings)
 `curl -X POST -H "Content-Type: application/json" -d '{"query": "generate something for this", "context": "asdasdasd"}' -H "Authorization: Bearer $(gcloud auth print-identity-token)" https://lit-search-model-t6p37v2uia-uw.a.run.app`

To deploy a microservice, run `./deploy_model.sh <service name> <huggingface model> <model type>`, e.g. `./deploy_model.sh reader-model deepset/electra-base-squad2 pipeline`

mruwnik commented on model_server/src/model_server/app.py:
```diff
@@ -0,0 +1,52 @@
+
+
+routes = [
+    Route("/", handler(DEFAULT), methods=["POST"]),
```
this will return the default action for the model

mruwnik replied:
yup

mruwnik replied:
actually, now I think about it, the default action should be configurable... Will add it

mruwnik commented on model_server/src/model_server/app.py:
```diff
@@ -0,0 +1,52 @@
+
+routes = [
+    Route("/", handler(DEFAULT), methods=["POST"]),
+    Route("/encoding", handler(ENCODE), methods=["POST"]),
```
These more specific endpoints will cause an error (returns a 400) if they can't support this call.
Now I think about it, it might be better to just have the default endpoint? How many models are there that can support both kinds of actions?

ccstan99 commented on model_server/src/model_server/settings.py:
```diff
@@ -0,0 +1,32 @@
+        raise Exception(
+            f'No model tokenizer provided! MODEL_TOKENIZER must point to a valid path: {MODEL_PATH.absolute()}'
+        )
+    if not PIPELINE_TYPE:
```
Were you running into problems with initializing the tokenizer? This shouldn't be an issue unique to pipeline though.

mruwnik replied:
no, that seemed to work. Though I haven't tried integrating this with pinecone, so maybe it's producing total garbage... I tried to keep it as close to the original code as possible

ccstan99 replied:
According the hugging face documentation, all of the language models are *supposed* to provide their own tokenizer. So if it's missing, there's something much more fundamentally wrong.

ccstan99 commented on model_server/src/model_server/settings.py:
```diff
@@ -0,0 +1,32 @@
+    if not PIPELINE_TYPE:
+        raise Exception('No PIPELINE_TYPE provided')
+elif MODEL_TYPE == ENCODER:
+    pass
```
The naming convention/options for MODEL_TYPE being PIPELINE and ENCODER feels a bit misleading. Functionally this should be fine for now. Let me think about this a bit more and I might rename later.

mruwnik replied:
renamed/removed

ccstan99 commented:
This might just be a documentation detail:

- reader model - returns answer given a question & context ... **not for duplicates**. I followed the naming convention from research papers. Maybe it'll be clearer to rename this to `qa_model` like this hugging face example below from https://huggingface.co/tasks/question-answering?
```
question = "Where do I live?"
context = "My name is Merve and I live in İstanbul."
qa_model(question = question, context = context)
## {'answer': 'İstanbul', 'end': 39, 'score': 0.953, 'start': 31}
```
- retriever model - returns encodings given text or list of text. **Additionally can return duplicates `paraphrase_mining` when given a list of strings/titles/questions.**

- literature search - returns encodings. We won't use `paraphrase_mining` here but it should be supported by the model so no harm keeping it.

ccstan99 commented on model_server/README.md:
```diff
@@ -0,0 +1,41 @@
+* `MODEL_TYPE` - one of `encoder` (used to generate encodings) or `pipeline` (for search)
+* `MODEL_PATH` - the path to the model. If not provided will be assumed to be `./models/<MODEL_NAME>`
+* `MODEL_TOKENIZER` - the path to the tokenizer. If not provided will assume it's the same as `MODEL_PATH`
+* `DEBUG` - debug mode - (True/False) - by default disabled
```
Yes, the tokenizer should be the same as MODEL_PATH by default in all of our cases.

mruwnik commented on model_server/pyproject.toml:
```diff
@@ -0,0 +1,32 @@
+license = {text = "MIT"}
+dependencies = [
+    "starlette",
+    "asyncio",
```
fastapi uses starlette under the hood. I went with starlette because it can handle async a lot better than flask can. It should be totally fine for only this to use starlette, as it's supposed to be a very simple service.
I'm still somewhat suspicious of flask. It's grown and matured a lot over the years, but it started out as an [april fools joke](https://lucumr.pocoo.org/2010/4/3/april-1st-post-mortem/). I don't want to bash it though - it's a lot nicer than say Django

ccstan99 replied:
Interesting! Thanks for sharing.

mruwnik commented on model_server/pyproject.toml:
```diff
@@ -0,0 +1,32 @@
+    "sentence-transformers",
+    "torch",
+    "transformers",
+]
```
I considered splitting this even further, to have different dependencies, but ultimately decided to keep it simple. It's possible to do magic with the `MODEL_TYPE` setting so it autoinstalls the appropriate dependencies. The actual library only weighs around 1MB, so that shouldn't be a concern

ccstan99 replied:
OK, sounds good.

mruwnik commented on deploy_model.sh:
```diff
@@ -0,0 +1,57 @@
+--min-instances=0 --memory 4G --cpu=2 --platform managed --no-traffic --tag=test \
+--service-account=service@stampy-nlp.iam.gserviceaccount.com \
+--update-secrets=PINECONE_API_KEY=PINECONE_API_KEY:latest,HUGGINGFACE_API_KEY=HUGGINGFACE_API_KEY:latest,CODA_TOKEN=CODA_TOKEN:latest,AUTH_PASSWORD=AUTH_PASSWORD:latest
+
```
very good catch

mruwnik commented:
> Thanks! These might just be a documentation detail but I want to clarify:
> 
>     * reader model - returns answer given a question & context ... **not for duplicates**. I followed the naming convention from research papers. Maybe it'll be clearer to rename this to `qa_model` like this hugging face example below from https://huggingface.co/tasks/question-answering?
> 
> 
> ```
> question = "Where do I live?"
> context = "My name is Merve and I live in İstanbul."
> qa_model(question = question, context = context)
> ## {'answer': 'İstanbul', 'end': 39, 'score': 0.953, 'start': 31}
> ```
> 
>     * retriever model - returns encodings given text or list of text. **Additionally can return duplicates `paraphrase_mining` when given a list of strings/titles/questions.**
> 
>     * literature search - returns encodings. We won't use `paraphrase_mining` here but it should be supported by the model so no harm keeping it.

That does make things a lot clearer :D I changed the reader-model to qa-model and added a question_answering endpoint for encoders

ccstan99 commented on model_server/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+# Fetch the appropriate model
+RUN mkdir models
+RUN git clone https://huggingface.co/${MODEL_NAME} ./models/${MODEL_NAME}
+RUN cd ./models/${MODEL_NAME} && git lfs install && git lfs pull
```
Does this order work? I thought `git lfs install` needs to run BEFORE `git clone` with the large huggingface models.

mruwnik replied:
it seems to work... I can move it over. Or perhaps the `apt-get install` also runs `git lfs install`? Interesting

ccstan99 replied:
Oh, I see you already have `apt-get install -y git-lfs` above. Then is `git lfs install` here redundant?

ccstan99 commented on model_server/src/model_server/app.py:
```diff
@@ -0,0 +1,49 @@
+    Route("/encoding", handler(settings.ENCODE), methods=["POST"]),
+    Route("/question_answering", handler(settings.QUESTION_ANSWERING), methods=["POST"]),
+    Route("/paraphrase_mining", handler(settings.PARAPHRASE_MINING), methods=["POST"]),
+]
```
It looks like all models expose 4 endpoints: default, encoding, question_answering and paraphrase_mining? It feels a bit more complicated than necessary but maybe it's clearer to have explicit endpoints rather than only defaults? Since it's for our internal use that's probably ok. Just to be clear:
- Encoders (retriever & lit) by default encode but can also do paraphrase_mining. 
- qa_model ONLY supports question_answering.

> That does make things a lot clearer :D I changed the reader-model to qa-model and added a question_answering endpoint for encoders

Encoders don't support question_answering but rather paraphrase_mining endpoint

mruwnik replied:
hmm. An alternative would be to have the routes list get created on the basis of the model type? Or have a sort of plugin type thingy, where each model type has its own module (or class) and depending on what MODEL_TYPE is set, the appropriate module gets used... 

Yup, paraphrase_mining. I'll get the names right someday :/

ccstan99 replied:
I think customizing by MODEL_TYPE would be better. The encoder(s) and qa_models are actually fundamentally quite different. Exposing the same endpoints doesn't "feel" right.

And I totally get the feeling. After a while, everything sounds like blah blah blah...

mruwnik replied:
Totally overhauled the code - now there is a per model type module which will encapsulate the whole functionality of that model

ccstan99 replied:
Great!

mruwnik commented on model_server/src/model_server/model_types/base_model.py:
```diff
@@ -0,0 +1,46 @@
+from model_server import settings
+
+
+def verify_fields_exist(*fields):
```
a bit of magic which allows one do things like:
```
class Bla:
     
     @verify_fields_exist('cats', 'dogs')
     def do_smth(self, payload):
        print(payload['cats'], payload['dogs'])

ble = Bla()

# These will work fine
ble.do_smth({'cats': 123, 'dogs': 42})
ble.do_smth({'cats': 123, 'dogs': 42, 'dolphins': 312})

# These will return an error object, as they're missing required keys
ble.do_smth({'dolphins': 312})
ble.do_smth({'dogs': 42, 'dolphins': 312})
ble.do_smth({'cats': 123, 'dolphins': 312})

# These will also fail, as the key is present, but the value is falsey

ble.do_smth({'cats': 123, 'dogs': None})
ble.do_smth({'cats': 123, 'dogs': ''})
ble.do_smth({'cats': 123, 'dogs': 0})
```

mruwnik commented on model_server/src/model_server/model_types/base_model.py:
```diff
@@ -0,0 +1,46 @@
+    return verifier_wrapper
+
+
+def handler(fn):
```
This is a decorator that will make sure that the request has a JSON object, extract it, and pass it to the wrapped function. So you can do something like:

```
class Bla:

    @handler
    def adder(payload):
        return payload.get('a') + payload.get('b')

def make_request(json):
    # create a proper request object with a `json()` function that will return the provided json
   
ble = Bla()
ble.adder(make_request(json={'a': 2, 'b': 4}))
```

ccstan99 reviewed (changes requested):
Looks good. Some renaming notes:
1. Make encoding the default action for encoders. 
2. Rename variables & filenames called PIPELINE to only question_answering
3. MODEL_TYPE options are 'encoder' and 'question_answering' with associated DEFAULT_ACTION which cannot be set by user. 

Let's use either MODEL_TYPE or ACTION but not both? Maybe pick ('encoder', 'encode' or 'encoding') and ('question_answering)?

ccstan99 reviewed (approved):
Let's not worry about renaming things. Since this code works, we can do a separate later to remove the pipeline/qa_model and replace it with the open Q&A using OpenAI API.

mruwnik commented on model_server/src/model_server/settings.py:
```diff
@@ -0,0 +1,41 @@
+    consts.PIPELINE: consts.QUESTION_ANSWERING,
+    consts.ENCODER: consts.ENCODE,
+}
+
```
updated

mruwnik commented on model_server/src/model_server/model_types/encoder.py:
```diff
@@ -0,0 +1,37 @@
+        elif self.default_action == PARAPHRASE_MINING:
+            return await self.paraphrase_mining(request)
+        else:
+            return self.error('No valid default action provided')
```
done

mruwnik commented on deploy_model.sh:
```diff
@@ -0,0 +1,60 @@
+       --build-arg MODEL_NAME=$MODEL_NAME \
+       --build-arg MODEL_TYPE=$MODEL_TYPE \
+       --build-arg DEFAULT_ACTION=$DEFAULT_ACTION
+docker push $IMAGE:latest
```
changed to "encoding"/"question-answering"
